File: resources/ios/Sources/PushFunctions.swift
import Foundation
import UIKit
import UserNotifications
import FirebaseMessaging

enum PushFunctions {

    // PushNotification.CheckPermission — no prompt.
    class CheckPermission: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let semaphore = DispatchSemaphore(value: 0)
            var status = "unknown"
            UNUserNotificationCenter.current().getNotificationSettings { settings in
                switch settings.authorizationStatus {
                case .authorized:    status = "granted"
                case .denied:        status = "denied"
                case .notDetermined: status = "not_determined"
                case .provisional:   status = "provisional"
                case .ephemeral:     status = "ephemeral"
                @unknown default:    status = "unknown"
                }
                semaphore.signal()
            }
            _ = semaphore.wait(timeout: .now() + 2)
            return BridgeResponse.success(data: ["status": status])
        }
    }

    // PushNotification.RequestPermission — core's enroll() calls this with {id, event}.
    // NOTE: `event` here comes from your own JS/PHP code, not a push payload —
    // lower risk than the incoming-message path, but PushObserver.dispatchToPHP()
    // validates it against the same FQCN pattern regardless (defense in depth).
    class RequestPermission: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            if let id = parameters["id"] as? String {
                PushObserver.shared.enrollmentId = id
            }
            if let event = parameters["event"] as? String, !event.isEmpty {
                PushObserver.shared.tokenEventClass = event
            }

            UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, _ in
                guard granted else { return }
                DispatchQueue.main.async {
                    UIApplication.shared.registerForRemoteNotifications()
                }
            }
            return BridgeResponse.success(data: ["requested": true])
        }
    }

    // PushNotification.GetToken
    class GetToken: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            if let cached = PushObserver.shared.cachedToken {
                return BridgeResponse.success(data: ["token": cached])
            }
            let semaphore = DispatchSemaphore(value: 0)
            var token = ""
            Messaging.messaging().token { result, _ in
                token = result ?? ""
                semaphore.signal()
            }
            _ = semaphore.wait(timeout: .now() + 3)
            return BridgeResponse.success(data: ["token": token])
        }
    }

    // PushNotification.ClearBadge
    class ClearBadge: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            DispatchQueue.main.async {
                if #available(iOS 16.0, *) {
                    UNUserNotificationCenter.current().setBadgeCount(0)
                } else {
                    UIApplication.shared.applicationIconBadgeNumber = 0
                }
                UNUserNotificationCenter.current().removeAllDeliveredNotifications()
            }
            return BridgeResponse.success(data: ["cleared": true])
        }
    }
}

File: resources/android/src/PushFunctions.kt
package com.iamdevroyal.push

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.google.firebase.messaging.FirebaseMessaging
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import com.nativephp.mobile.ui.MainActivity
import org.json.JSONObject

object PushFunctions {

    // PushNotification.CheckPermission
    class CheckPermission(@Suppress("UNUSED_PARAMETER") activity: FragmentActivity? = null) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val context = PushRuntime.appContext
                ?: return BridgeResponse.success(mapOf("status" to "unknown"))

            val status = when {
                Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU ->
                    if (NotificationManagerCompat.from(context).areNotificationsEnabled()) "granted" else "denied"
                ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) ==
                    PackageManager.PERMISSION_GRANTED -> "granted"
                else -> "not_determined"
            }
            return BridgeResponse.success(mapOf("status" to status))
        }
    }

    // PushNotification.RequestPermission — core's enroll() passes {id, event}.
    // NOTE: the `event` override here comes from your own JS/PHP code calling
    // enroll(), not from a push payload — lower risk than the incoming-message
    // path, but PushDispatch.dispatch() validates it against the same FQCN
    // pattern regardless (defense in depth; a compromised WebView could call
    // this bridge function directly too).
    class RequestPermission(@Suppress("UNUSED_PARAMETER") activity: FragmentActivity? = null) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            (parameters["id"] as? String)?.let { PushRuntime.enrollmentId = it }
            (parameters["event"] as? String)?.takeIf { it.isNotEmpty() }?.let { PushRuntime.tokenEventClass = it }

            MainActivity.instance?.let { activity ->
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
                    ContextCompat.checkSelfPermission(activity, Manifest.permission.POST_NOTIFICATIONS) !=
                    PackageManager.PERMISSION_GRANTED
                ) {
                    ActivityCompat.requestPermissions(
                        activity, arrayOf(Manifest.permission.POST_NOTIFICATIONS), 4711
                    )
                }
            }

            FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
                if (task.isSuccessful) task.result?.let { token ->
                    PushRuntime.cachedToken = token
                    PushDispatch.dispatch(
                        PushRuntime.tokenEventClass,
                        JSONObject().put("token", token).put("id", PushRuntime.enrollmentId)
                    )
                }
            }
            return BridgeResponse.success(mapOf("requested" to true))
        }
    }

    // PushNotification.GetToken
    class GetToken(@Suppress("UNUSED_PARAMETER") activity: FragmentActivity? = null) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            PushRuntime.cachedToken?.let { return BridgeResponse.success(mapOf("token" to it)) }

            val latch = java.util.concurrent.CountDownLatch(1)
            var token = ""
            FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
                if (task.isSuccessful) token = task.result ?: ""
                latch.countDown()
            }
            latch.await(3, java.util.concurrent.TimeUnit.SECONDS)
            if (token.isNotEmpty()) PushRuntime.cachedToken = token
            return BridgeResponse.success(mapOf("token" to token))
        }
    }

    // PushNotification.ClearBadge
    class ClearBadge(@Suppress("UNUSED_PARAMETER") activity: FragmentActivity? = null) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            PushRuntime.appContext?.let { NotificationManagerCompat.from(it).cancelAll() }
            return BridgeResponse.success(mapOf("cleared" to true))
        }
    }
}
